cartographer/mapping_3d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -150,43 +144,37 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
 
   // Set the starting point.
   CHECK(!submap_data_.empty());
-  CHECK(!submap_data_[0].empty());
-  // TODO(hrapp): Move ceres data into SubmapData.
-  std::vector<std::map<int, CeresPose>> C_submaps(submap_data_.size());
+  CHECK(submap_data_.Contains(mapping::SubmapId{0, 0}));
+  mapping::MapById<mapping::SubmapId, CeresPose> C_submaps;
   std::vector<std::map<int, CeresPose>> C_nodes(node_data_.size());
   bool first_submap = true;
-  for (size_t trajectory_id = 0; trajectory_id != submap_data_.size();
-       ++trajectory_id) {
-    const bool frozen = frozen_trajectories.count(trajectory_id) != 0;
-    for (const auto& index_submap_data : submap_data_[trajectory_id]) {
-      const int submap_index = index_submap_data.first;
-      if (first_submap) {
-        first_submap = false;
-        // Fix the first submap of the first trajectory except for allowing
-        // gravity alignment.
-        C_submaps[trajectory_id].emplace(
-            std::piecewise_construct, std::forward_as_tuple(submap_index),
-            std::forward_as_tuple(
-                index_submap_data.second.pose, translation_parameterization(),
-                common::make_unique<ceres::AutoDiffLocalParameterization<
-                    ConstantYawQuaternionPlus, 4, 2>>(),
-                &problem));
-        problem.SetParameterBlockConstant(
-            C_submaps[trajectory_id].at(submap_index).translation());
-      } else {
-        C_submaps[trajectory_id].emplace(
-            std::piecewise_construct, std::forward_as_tuple(submap_index),
-            std::forward_as_tuple(
-                index_submap_data.second.pose, translation_parameterization(),
-                common::make_unique<ceres::QuaternionParameterization>(),
-                &problem));
-      }
-      if (frozen) {
-        problem.SetParameterBlockConstant(
-            C_submaps[trajectory_id].at(submap_index).rotation());
-        problem.SetParameterBlockConstant(
-            C_submaps[trajectory_id].at(submap_index).translation());
-      }
+  for (const auto& submap_id_data : submap_data_) {
+    const bool frozen =
+        frozen_trajectories.count(submap_id_data.id.trajectory_id) != 0;
+    if (first_submap) {
+      first_submap = false;
+      // Fix the first submap of the first trajectory except for allowing
+      // gravity alignment.
+      C_submaps.Insert(
```

Comment:
Why not add an Emplace() to MapById?

Reply:
Unfortunately it seems not trivial. With C++11 forwarding a tuple with the arguments to a constructor is complicated, it gets easier with C++14 and easy to do in C++17. Changing the `CeresPose` to a single argument being a tuple seems also less than ideal. Leaving it for now as discussed.